### PR TITLE
fix: resolve issues sqlite can use timestamp type column

### LINKED ISSUE #10140: timestamp type column doesn't work when using sqlite
### Issue description

timestamp type column doesn't work when using sqlite

### Expected Behavior

``` 
@Entity()
export class Photo {
    @PrimaryGeneratedColumn({
        type: 'int',
    })
    id: number

    @Column('varchar')
    name: string;

    @Column('timestamp')
    publishedAt: Date
}
```
both mysql-family and sqlite, entity works correctly.

### Actual Behavior

DataTypeNotSupportedError: Data type "timestamp" in "table.column" is not supported by "sqlite" database.

### Steps to reproduce

``` 
@Entity()
export class Photo {
    @PrimaryGeneratedColumn({
        type: 'int',
    })
    id: number

    @Column('varchar')
    name: string;

    @Column('timestamp')
    publishedAt: Date
}
```
just use column decorator using timestamp type.

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    | mac         |
| Node.js version     | 16.x  |
| Typescript version  | 5.x  |
| TypeORM version     | 0.3.4  |

### Additional Context

_No response_

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [ ] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [X] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, and I know how to start.

### PR DESCRIPTION
**Pull request by bejewelkyoungminkim**
Closes: #10140

### Description of change

I'm using typeorm with mysql, all of my code works based on typeorm. 
But when I tried to converting my database to sqlite, I found that with sqlite, timestamp column doesn't work.

```
DataTypeNotSupportedError: Data type "timestamp" in "xxxx" is not supported by "sqlite" database.
```

In this situation I have to change all of my code that using timestamp column. that's why I made this PR.
I made timestamp type work using bigint type with sqlite.

If you have any good idea or advice, let me know.
I'll do my best to fix this issue.
thank you.

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #10140`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: test/github-issues/10140/issue-10140.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+describe("github issues > #10140 timestamp type column doesn't work when using sqlite", () => {
+    let dataSources: DataSource[]
+    before(
+        async () =>
+            (dataSources = await createTestingConnections({
+                entities: [__dirname + "/entity/*{.js,.ts}"],
+                schemaCreate: true,
+                dropSchema: true,
+            })),
```
Comment (kusigit) [inline test/github-issues/10140/issue-10140.ts:19]: It's not necessary to load the other drivers.
```suggestion
                enabledDrivers: ["sqlite"],
            })),
```

**Comment by bejewelkyoungminkim** [comment]
@kusigit thanks : )

**Review by hinogi — APPROVED** [review]
LGTM

**Comment by t-higuchi** [comment]
Could you give us an update on this?
If circleci error prevent this PR from merge, please relaunch it.  #10101 should fix the problem ..

**Comment by bejewelkyoungminkim** [comment]
@t-higuchi 
I'm waiting for merging this PR.
CircleCi shows error about oracle. But It is not releated this PR

**Comment by jasl** [comment]
> @t-higuchi I'm waiting for merging this PR. CircleCi shows error about oracle. But It is not releated this PR

I guess his meaning is rebase the master branch then the CI will pass

**Comment by bejewelkyoungminkim** [comment]
@jasl 
Oh, I see. I have merged master branch into my branch.
all of ci are passed : )
Thank you.

**Comment by bejewelkyoungminkim** [comment]
I'm looking forward to merging my PR. I have been waiting for 4 months. 
somebody help me.
@hinogi @kusigit

**Comment by jasl** [comment]
> @jasl Oh, I see. I have merged master branch into my branch. all of ci are passed : ) Thank you.

Thank you for your contribution! I hope this can be merged ASAP, and make a release :)

**Comment by kala888** [comment]
did you released this feature?

**Comment by bejewelkyoungminkim** [comment]
> did you released this feature?

not yet. I'm looking forward to releasing this.

**Comment by abeebola** [comment]
@bejewelkyoungminkim do you mind adding support for other timestamp types? For example, `timestamp with time zone`

**Comment by bejewelkyoungminkim** [comment]
> @bejewelkyoungminkim do you mind adding support for other timestamp types? For example, `timestamp with time zone`. Also, at the point of returning the data, can you convert back to `Date`?

`TIMESTAMP WITH TIME ZONE` this type must be for postgresql. But I'm not using postgresql. Somebody will help

**Comment by kala888** [comment]
> > @bejewelkyoungminkim do you mind adding support for other timestamp types? For example, `timestamp with time zone`. Also, at the point of returning the data, can you convert back to `Date`?
> 
> `TIMESTAMP WITH TIME ZONE` this type must be for postgresql. But I'm not using postgresql. Somebody will help

I'm using SQLite, and my temporary solution is to create a transformer at column definition.

```javascript
  @Column({ type: 'datetime', transformer: new LocalDateTransformer() })
  createTime: Date;
```

local-date-transformer.ts
```javascript
import { ValueTransformer } from 'typeorm';
import * as moment from 'moment';

// 搞不定sqlite时区，暂时用这个转换，存到数据库里面+8小时
export class LocalDateTransformer implements ValueTransformer {

  to(date?: Date): string | null {
    return moment(date).add(8, 'h').format('YYYY-MM-DD HH:mm:ss');
  }

  from(date?: string): Date | null {
    if (date) {
      try {
        return moment(date).toDate();
      } catch (e) {
        console.log('LocalDateTransformer error', e);
      }
      return null;
    }
  }
}

```

**Comment by abeebola** [comment]
> > > @bejewelkyoungminkim do you mind adding support for other timestamp types? For example, `timestamp with time zone`. Also, at the point of returning the data, can you convert back to `Date`?
> > 
> > 
> > `TIMESTAMP WITH TIME ZONE` this type must be for postgresql. But I'm not using postgresql. Somebody will help
> 
> I'm using SQLite, and my temporary solution is to create a transformer at column definition.
> 
> ```js
>   @Column({ type: 'datetime', transformer: new LocalDateTransformer() })
>   createTime: Date;
> ```
> 
> local-date-transformer.ts
> 
> ```js
> import { ValueTransformer } from 'typeorm';
> import * as moment from 'moment';
> 
> // 搞不定sqlite时区，暂时用这个转换，存到数据库里面+8小时
> export class LocalDateTransformer implements ValueTransformer {
> 
>   to(date?: Date): string | null {
>     return moment(date).add(8, 'h').format('YYYY-MM-DD HH:mm:ss');
>   }
> 
>   from(date?: string): Date | null {
>     if (date) {
>       try {
>         return moment(date).toDate();
>       } catch (e) {
>         console.log('LocalDateTransformer error', e);
>       }
>       return null;
>     }
>   }
> }
> ```

Thanks for the code snippet @kala888 . Unfortunately, I only want to use SQLite for running tests and this approach won't work.

**Comment by pleerock** [comment]
According to https://www.sqlite.org/datatype3.html SQLite doesn't support timestamp data type. We must only have types supported by underlying database to prevent confusion / possible future types added by the database driver itself.

**Comment by bejewelkyoungminkim** [comment]
> According to https://www.sqlite.org/datatype3.html SQLite doesn't support timestamp data type. We must only have types supported by underlying database to prevent confusion / possible future types added by the database driver itself.

Various databases have various types.
They are not always using same name for similar or same type.
That's why we need to normalize column type.
And typeorm already provide a way to normalize columns in AbstractSqliteDriver.ts


```
    normalizeType(column: {
        type?: ColumnType
        length?: number | string
        precision?: number | null
        scale?: number
    }): string {
        if (column.type === Number || column.type === "int") {
            return "integer"
        } else if (column.type === String) {
            return "varchar"
        } else if (column.type === Date) {
            return "datetime"
        } else if (column.type === Boolean) {
            return "boolean"
        } else if (column.type === "uuid") {
            return "varchar"
        } else if (column.type === "simple-array") {
            return "text"
        } else if (column.type === "simple-json") {
            return "text"
        } else if (column.type === "simple-enum") {
            return "varchar"
        } else {
            return (column.type as string) || ""
        }
    }
```

**Comment by pleerock** [comment]
> Various databases have various types.

Correct, different databases support different types. However, it's up to you how exactly you want to "normalize" the types. Because once you start to "normalize" types manually different use cases and edge cases start to arise and it brings lot of problems. For example, why "timestamp" leads to "bigint"? It's inefficient, let's use some different more effecient type. Some people can complain about it and say "no, we need exactly bigint because it supports numeric operations". Others can say we need to have timestamps as strings to support some string-based operations. This is just a primitive example, variety of issues can be much larger. That's why we keep it up to user to define type normalization. 

> And typeorm already provide a way to normalize columns 

Yes, we do provide opinion-based normalization for JavaScript types, since we support column definition without explicit types defined and set of `simple-*` types as an exception. We don't have plans to extend this list.

**Comment by jasl** [comment]
> > Various databases have various types.
> 
> Correct, different databases support different types. However, it's up to you how exactly you want to "normalize" the types. Because once you start to "normalize" types manually different use cases and edge cases start to arise and it brings lot of problems. For example, why "timestamp" leads to "bigint"? It's inefficient, let's use some different more effecient type. Some people can complain about it and say "no, we need exactly bigint because it supports numeric operations". Others can say we need to have timestamps as strings to support some string-based operations. This is just a primitive example, variety of issues can be much larger. That's why we keep it up to user to define type normalization.
> 
> > And typeorm already provide a way to normalize columns
> 
> Yes, we do provide opinion-based normalization for JavaScript types, since we support column definition without explicit types defined and set of `simple-*` types as an exception. We don't have plans to extend this list.



> According to https://www.sqlite.org/datatype3.html SQLite doesn't support timestamp data type. We must only have types supported by underlying database to prevent confusion / possible future types added by the database driver itself.

Sqlite uses Integer to represent Timestamp

```
INTEGER as Unix Time, the number of seconds since 1970-01-01 00:00:00 UTC.
```

So this normalization is valid

**Comment by abeebola** [comment]
> > Various databases have various types.
> 
> Correct, different databases support different types. However, it's up to you how exactly you want to "normalize" the types. Because once you start to "normalize" types manually different use cases and edge cases start to arise and it brings lot of problems. For example, why "timestamp" leads to "bigint"? It's inefficient, let's use some different more effecient type. Some people can complain about it and say "no, we need exactly bigint because it supports numeric operations". Others can say we need to have timestamps as strings to support some string-based operations. This is just a primitive example, variety of issues can be much larger. That's why we keep it up to user to define type normalization.
> 
> > And typeorm already provide a way to normalize columns
> 
> Yes, we do provide opinion-based normalization for JavaScript types, since we support column definition without explicit types defined and set of `simple-*` types as an exception. We don't have plans to extend this list.

Can we at least have a way to set this up once instead of having to use a transformer for each column in each entity? This would help for situations where you want to run integration tests with SQLite when your production app uses a standard RDBMS like Postgres etc..

**Comment by pleerock** [comment]
> Sqlite uses Integer to represent Timestamp

we didn't get it merged yet, and there is already one thing to debate: integer vs bigint for sqlite timestamps

> So this normalization is valid, it's even more valid than normalizing uuid to varchar as typeorm already did

one another debate opened for uuid normalization implementation. 

See we already have complains about currently implemented normalization strategies.
That's why it's better to keep these decisions to users.

**Comment by pleerock** [comment]
> Can we at least have a way to set this up once instead of having to use a transformer for each column in each entity? This would help for situations where you want to run integration tests with SQLite when your production app uses a standard RDBMS like Postgres etc..

```ts
@Column({ 
   type: DATABASE_TYPE === "sqlite" ? "bigint" : "timestamp"
})
date: number | string
```

define DATABASE_TYPE somewhere in your code, or even in env variable for your particular application. There can be other ways of doing that, it's just one way here.

**Comment by bejewelkyoungminkim** [comment]
If not to merge this pr is majority, do we have any other solution for this? 
It is hard to find a convenient solution for nestjs+typeorm users.

**Comment by pleerock** [comment]
> but if DATABASE_TYPE is injected when app started, we can't use this way.

I don't understand what you mean here. Why you can't use it such way?

**Comment by bejewelkyoungminkim** [comment]
```js
// config 
const DATABASE_TYPE = 'sqlite';
```

```js
// entity

@Column({ 
   type: DATABASE_TYPE === "sqlite" ? "bigint" : "timestamp"
})
date: number | string
```

in this case, DATABASE_TYPE is already declared, we can do that.


but nestjs with configSevice

```js
TypeOrmModule.forRootAsync({
  imports: [ConfigModule],
  useFactory: (configService: ConfigService) => ({
    type: configService.get('DATABASE_TYPE'),
    host: configService.get('HOST'),
    port: +configService.get('PORT'),
    username: configService.get('USERNAME'),
    password: configService.get('PASSWORD'),
    database: configService.get('DATABASE'),
    entities: [...],
  }),
  inject: [ConfigService],
});
```

```js
//  ❌  @Column decorator cannot use configService

@Column({ 
   type: configService.get('DATABASE_TYPE') === "sqlite" ? "bigint" : "timestamp"
})
date: number | string
```

right?

**Comment by pleerock** [comment]
This is specific of your application architecture. If you decide to get value from some `configService` -> you should be able to understand how to access it decorators. If that's something that is not accessible from the root of your file -> you need to take values from the other source / place. Make things simpler and everything will work :)

**Comment by bejewelkyoungminkim** [comment]
You can find various ways to make it possible. Yes, I know.
But we have to make lots of decorators with condition or transformers.
I think my solution is best way to prepare hydrated value for "Date" type and timstamp column.

**Comment by pleerock** [comment]
@bejewelkyoungminkim I'm sorry, but this is too much opinionated and we don't want to go such way and have multiple other issues of users complaining about specific behaviors of "normalized opinionated types".

If you have too many conditions in your decorators - consider creating your own decorators and re-use them across your codebase.

**Comment by johngrimsey** [comment]
```typescript
...
// in entity definition
@Column({ type: columnTypeTimestamp() })
columnName: Date
...

// utils/db/column-type-timestamp.ts
function columnTypeTimestamp() {
   return useSqlite() ? 'bigint' : 'timestamp'
}

// utils/db/column-type-json.ts
function columnTypeJson() {
    return useSqlite() ? 'simple-json' : 'jsonb'
}

// utils/env/use-sqlite.ts
function useSqlite() {
  return getNodeEnv() === 'test'
}

// utils/env/get-node-env.ts
function getNodeEnv() {
  return process.env['NODE_ENV']
}
```

**Pull request closed without merging**